go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,605 @@
+				"err", err,
+				"stack", string(callStack))
+
+			// remove this goroutine from in-memory state
+			if err := s.inMemoryState.RemoveLogItem(runUID); err != nil {
+				s.Logger.WarnWithCtx(ctx, "Failed to remove item from in memory state")
+			}
+
+			// cancel all other goroutines spawned from this one
+			cancelCtxFunc()
+		}
+	}()
+
+	// signal "main" function that goroutine is up
+	startedStreamingGoroutine <- true
+
+	// create a log file to the pod
+	logFilePath := s.resolvePodLogFilePath(runUID, podName)
+	if err := common.EnsureFileExists(logFilePath); err != nil {
+		s.Logger.ErrorWithCtx(ctx,
+			"Failed to ensure log file",
+			"runUID", runUID,
+			"logFilePath", logFilePath)
+		return
+	}
+
+	// get logs from pod, and keep the stream open (follow)
+	podLogOptions := &v1.PodLogOptions{
+		Follow: true,
+	}
+	restClientRequest := s.kubeClientSet.CoreV1().Pods(s.namespace).GetLogs(podName, podLogOptions)
+
+	// initialize stream and error for the while loop
+	var (
+		streamErr error
+		stream    io.ReadCloser
+	)
+
+	// stream logs - retry if failed
+	for {
+		stream, streamErr = restClientRequest.Stream(ctx)
+		if streamErr == nil {
+			break
+		}
+		s.Logger.WarnWithCtx(ctx, "Failed to get pod log stream read/closer, retrying", "runUID", runUID, "err", streamErr.Error())
+		time.Sleep(1 * time.Second)
+	}
+	defer stream.Close() // nolint: errcheck
+
+	for {
+		keepLogging, err := s.streamPodLogs(ctx, runUID, logFilePath, stream)
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "An error occurred while streaming pod logs", "err", err)
+		}
+		if keepLogging {
+			continue
+		}
+		break
+	}
+
+	s.Logger.DebugWithCtx(ctx,
+		"Removing item from state file",
+		"runUID", runUID,
+		"podName", podName)
+
+	// remove run from state file
+	if err := s.stateStore.RemoveLogItem(runUID); err != nil {
+		s.Logger.WarnWithCtx(ctx, "Failed to remove log item from state file")
+	}
+}
+
+func (s *LogCollectorServer) streamPodLogs(ctx context.Context,
+	runUID,
+	logFilePath string,
+	stream io.ReadCloser) (bool, error) {
+
+	// create a reader from the stream, to allow peeking into it
+	streamReader := bufio.NewReader(stream)
+
+	// wait for the stream ro have logs before reading them
+	if !s.hasLogs(ctx, runUID, streamReader) {
+		s.Logger.WarnWithCtx(ctx, "Stream doesn't have logs or context has been canceled", "runUID", runUID)
+		return false, nil
+	}
+
+	// open log file in read/write and append, to allow reading the logs while we write more logs to it
+	openFlags := os.O_RDWR | os.O_APPEND
+	file, err := os.OpenFile(logFilePath, openFlags, 0644)
+	if err != nil {
+		s.Logger.WarnWithCtx(ctx, "Failed to open file", "err", err, "logFilePath", logFilePath)
+		return true, err
+	}
+	defer file.Close() // nolint: errcheck
+
+	// spin a goroutine that will unblock `CopyBuffer` if context is dead
+	copyBufferDone := make(chan struct{}, 1)
+	go s.cancelOnContext(ctx, file, copyBufferDone)
```

Comment:
i think it is better to do 
```
go func() {
}()
```

and put the `cancelOnContext` file contents inline for readability.